INTMDB-364 Add support for serverless private endpoints (AWS + Azure) (#913)
mongodb/terraform-provider-mongodbatlas: website/docs/r/privatelink_endpoint_service_serverless.html.markdown

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: privatelink_endpoint_service_serverless"
sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoint-service-serverless"
description: |-
Describes a Serverless PrivateLink Endpoint Service
---

> # Data Source: privatelink_endpoint_service_serverless

Reviewer: ```suggestion
# Resource: privatelink_endpoint_service_serverless
```

`privatelink_endpoint_service_serverless` Provides a Serverless PrivateLink Endpoint Service resource.
This is the second of two resources required to configure PrivateLink for Serverless, the first is [mongodbatlas_privatelink_endpoint_serverless](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_serverless).

> -> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

Reviewer: ```suggestion
-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

-> **NOTE:** Feature unavailable in Free and Shared-Tier Clusters
This feature is not available for M0 free clusters, M2, and M5 clusters. To learn more about which features are unavailable, see [Atlas M0 (Free Cluster), M2, and M5 Limitations.](https://www.mongodb.com/docs/atlas/reference/free-shared-limitations/#std-label-atlas-free-tier)
```

Reviewer: @martinstibbe delete this one

  resource_group_name = var.resource_group_name
  subnet_id           = azurerm_subnet.test.id
  private_service_connection {
    name                           = mongodbatlas_privatelink_endpoint_serverless.test.private_link_service_name
    private_connection_resource_id = mongodbatlas_privatelink_endpoint_serverless.test.private_link_service_resource_id
    is_manual_connection           = true
    request_message                = "Azure Private Link test"
  }

}

resource "mongodbatlas_privatelink_endpoint_service_serverless" "test" {
  project_id                  = mongodbatlas_privatelink_endpoint_serverless.test.project_id
  instance_name               = mongodbatlas_serverless_instance.test.name
  endpoint_id                 = mongodbatlas_privatelink_endpoint_serverless.test.endpoint_id
  cloud_endpoint_id           = azurerm_private_endpoint.test.id 
  private_endpoint_ip_address = azurerm_private_endpoint.test.private_service_connection.0.private_ip_address
  provider_name               = "AZURE"
  comment                     = "test"
}
```

## Argument Reference

* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.

> * `instance_name` - (Required)Human-readable label that identifies the serverless instance.

Reviewer: ```suggestion
* `instance_name` - (Required) Human-readable label that identifies the serverless instance.
```

> * `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint's network interface. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.

Reviewer: ```suggestion
* `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint's network interface. Atlas supports AWS private endpoints using the [AWS PrivateLink](https://aws.amazon.com/privatelink/) feature.
```

* `comment` - (Optional) Human-readable string to associate with this private endpoint.

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

* `endpoint_service_name` - Unique string that identifies the PrivateLink endpoint service.
* `private_link_service_resource_id` - Root-relative path that identifies the Azure Private Link Service that MongoDB Cloud manages.
* `private_endpoint_ip_address` - IPv4 address of the private endpoint in your Azure VNet that someone added to this private endpoint service.
* `cloud_provider_endpoint_id` - Unique string that identifies the private endpoint's network interface.
* `comment` - Human-readable string to associate with this private endpoint.
* `error_message` - Human-readable error message that indicates the error condition associated with establishing the private endpoint connection.
* `status` - Human-readable label that indicates the current operating status of the private endpoint. Values include: RESERVATION_REQUESTED, RESERVED, INITIATING, AVAILABLE, FAILED, DELETING.

> For more information see: [MongoDB Atlas API - Serverless Private Endpoints](https://www.mongodb.com/docs/atlas/reference/api/serverless-private-endpoints-get-one/).

Reviewer: Missing IMPORT section

  name                = "endpoint-test"
  location            = data.azurerm_resource_group.test.location
  resource_group_name = var.resource_group_name
  subnet_id           = azurerm_subnet.test.id
  private_service_connection {
    name                           = mongodbatlas_privatelink_endpoint_serverless.test.private_link_service_name
    private_connection_resource_id = mongodbatlas_privatelink_endpoint_serverless.test.private_link_service_resource_id
    is_manual_connection           = true
    request_message                = "Azure Private Link test"
  }

}

resource "mongodbatlas_privatelink_endpoint_service_serverless" "test" {
  project_id                  = mongodbatlas_privatelink_endpoint_serverless.test.project_id
  instance_name               = mongodbatlas_serverless_instance.test.name
  endpoint_id                 = mongodbatlas_privatelink_endpoint_serverless.test.endpoint_id
  cloud_endpoint_id           = azurerm_private_endpoint.test.id 
  private_endpoint_ip_address = azurerm_private_endpoint.test.private_service_connection.0.private_ip_address
  provider_name               = "AZURE"
  comment                     = "test"

> }

Reviewer 2: The AWS example has the block for the actual serverless instance creation, please add it for the Azure example as well.  You have one in the data source so just use that (which corrected region).

```

## Argument Reference

* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.
* `instance_name` - (Required) Human-readable label that identifies the serverless instance.

> * `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint's network interface. Atlas supports AWS private endpoints using the [AWS PrivateLink](https://aws.amazon.com/privatelink/) feature.

Reviewer 2: Similar to above - remove AWS.  Also 24, not 22:
```suggestion
* `endpoint_id` - (Required) Unique 24-hexadecimal digit string that identifies the private endpoint.

```

* `cloud_provider_endpoint_id` - (Optional) Unique string that identifies the private endpoint's network interface.
* `private_endpoint_ip_address` - (Optional) IPv4 address of the private endpoint in your Azure VNet that someone added to this private endpoint service.
* `provider_name` - (Required) Cloud provider for which you want to create a private endpoint. Atlas accepts `AWS`, `AZURE`.
* `comment` - (Optional) Human-readable string to associate with this private endpoint.

> 

Reviewer 2: Since you added timeouts don't we need something like https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service#timeouts?

---
layout: "mongodbatlas"
page_title: "MongoDB Atlas: privatelink_endpoint_service_serverless"
sidebar_current: "docs-mongodbatlas-datasource-privatelink-endpoint-service-serverless"
description: |-

> Describes a Serverless PrivateLink endpoint service

Reviewer: ```suggestion
Describes a Serverless PrivateLink Endpoint Service
```

Author: Corrected

---

# Data Source: privatelink_endpoint_service_serverless

> `privatelink_endpoint_service_serverless` Provides a Serverless PrivateLink endpoint service resource.

Reviewer: ```suggestion
`privatelink_endpoint_service_serverless` Provides a Serverless PrivateLink Endpoint Service resource.
```

Author: Corrected

> `privatelink_endpoint_service_serverless` Provides a Serverless PrivateLink Endpoint Service resource.

Reviewer 2: Similar to above
```suggestion
`privatelink_endpoint_service_serverless` Provides a Serverless PrivateLink Endpoint Service resource.  This is the second of two resources required to configure PrivateLink for Serverless, the first is [mongodbatlas_privatelink_endpoint_serverless](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_serverless).  
```

-> **NOTE:** Groups and projects are synonymous terms. You may find group_id in the official documentation.

## Example Usage

### Basic
```terraform
data "mongodbatlas_privatelink_endpoint_service_serverless" "test" {
	project_id   = "<PROJECT_ID>"
	instance_name = mongodbatlas_serverless_instance.test.name
	endpoint_id = mongodbatlas_privatelink_endpoint_serverless.test.endpoint_id

> }

Reviewer 2: Did you mean to include this in the data source docs example perhaps?

resource "mongodbatlas_privatelink_endpoint_serverless" "test" {
	project_id   = "<PROJECT_ID>"
	instance_name = mongodbatlas_serverless_instance.test.name
	provider_name = "AWS"
}
	  

> 	  

Reviewer 2: As noted in the tutorial: https://www.mongodb.com/docs/atlas/security-serverless-private-endpoint/ there is a step between "mongodbatlas_privatelink_endpoint_serverless" and resource "mongodbatlas_privatelink_endpoint_service_serverless" which is to create the interface endpoint on the AWS side. We provide how to do this in the regular privatelink docs: the https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/resources/privatelink_endpoint_service Let's do the same here.

Reviewer 2: This is still missing the AWS block.  Azure has one so just need this like that.

resource "mongodbatlas_privatelink_endpoint_service_serverless" "test" {
	project_id   = "<PROJECT_ID>"

> 	instance_name = "test-db"

Reviewer 2: I don't think they'd want to hard code this but use mongodbatlas_serverless_instance.test.name again?

	endpoint_id = mongodbatlas_privatelink_endpoint_serverless.test.endpoint_id
	provider_name = "AWS"
	comment = "New serverless endpoint"
}

resource "mongodbatlas_serverless_instance" "test" {
	project_id   = "<PROJECT_ID>"
	name         = "test-db"
	provider_settings_backing_provider_name = "AWS"
	provider_settings_provider_name = "SERVERLESS"
	provider_settings_region_name = "US_EAST_1"
	continuous_backup_enabled = true
}
```

## Argument Reference

* `project_id` - (Required) Unique 24-digit hexadecimal string that identifies the project.

> * `instance_name` - (Required) Serverless instance name to attach private

Reviewer: ```suggestion
* `instance_name` - (Required) Human-readable label that identifies the serverless instance associated with the tenant endpoint
```

Author: Corrected

> * `instance_name` - (Required)Human-readable label that identifies the serverless instance associated with the tenant endpoint

Reviewer 2: The source docs actually don't make sense here externally :) 
```suggestion
* `instance_name` - (Required)Human-readable label that identifies the serverless instance 
```

> * `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.

Reviewer: ```suggestion
* `endpoint_id` - (Required) Unique 22-character alphanumeric string that identifies the private endpoint's network interface. Atlas supports AWS private endpoints using the [|aws| PrivateLink](https://aws.amazon.com/privatelink/) feature.
```

Author: Corrected

## Attributes Reference

In addition to all arguments above, the following attributes are exported:

> * `endpoint_service_name` - Unique string that identifies the PrivateLink endpoint service. MongoDB Cloud returns null while it creates the endpoint service.

Reviewer 2: Similar to above, do we need to mention the null? 
```suggestion
* `endpoint_service_name` - Unique string that identifies the PrivateLink endpoint service. 
```

* `cloud_provider_endpoint_id` - Unique string that identifies the private endpoint's network interface.
* `comment` - Human-readable string to associate with this private endpoint.
* `error_message` - Human-readable error message that indicates the error condition associated with establishing the private endpoint connection.
* `status` - Human-readable label that indicates the current operating status of the private endpoint. Values include: RESERVATION_REQUESTED, RESERVED, INITIATING, AVAILABLE, FAILED, DELETING.

> For more information see: [MongoDB Atlas API - Serverless Private Endpoints](https://www.mongodb.com/docs/atlas/reference/api/serverless-private-endpoints-get-one/)  and [MongoDB Atlas API - Online Archive](https://docs.atlas.mongodb.com/reference/api/online-archive/) Documentation.

Reviewer 2: ```suggestion
For more information see: [MongoDB Atlas API - Serverless Private Endpoints](https://www.mongodb.com/docs/atlas/reference/api/serverless-private-endpoints-get-one/).
```
